### PR TITLE
Fixing race condition in ConnectionListener and fixing unaligned acce…

### PR DESCRIPTION
**Pull request by MushMal**
…ss in h264 rtp depeyer

The refactoring of the ConnectionListener has an additional benefit of making the interlocking easier to understand. Potentially addressing issues:

Issue #995 and #1019 

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by hassanctech — COMMENTED** [review]
Bunch of tests/asserts are failing

File: src/source/Ice/SocketConnection.c [hunk 1]
```diff
@@ -35,6 +35,7 @@ STATUS createSocketConnection(KVS_IP_FAMILY_TYPE familyType, KVS_SOCKET_PROTOCOL
     }
     ATOMIC_STORE_BOOL(&pSocketConnection->connectionClosed, FALSE);
     ATOMIC_STORE_BOOL(&pSocketConnection->receiveData, FALSE);
+    ATOMIC_STORE(&pSocketConnection->inUse, FALSE);
```
Comment (hassanctech) [inline src/source/Ice/SocketConnection.c:38]: should use ATOMIC_STORE_BOOL here as well

Comment (MushMal) [reply src/source/Ice/SocketConnection.c:38]: D'oh. Will change.

BTW. It seems that we do have TSAN exception for this particular failure case but for some reason it doesn't get picked up. Need to re-run

File: tst/suppressions/TSAN.supp [hunk 2]
```diff
@@ -209,78 +110,7 @@ race:freeSignalingClient
...
-#     #25 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) <null> (webrtc_client_test+0x000000672819)
-#
-#   Mutex M721 acquired here while holding mutex M730 in thread T72:
-#     #0 pthread_mutex_lock <null> (webrtc_client_test+0x0000004df310)
-#     #1 defaultLockMutex /usr/src/amazon-kinesis-video-streams-webrtc-sdk-c/open-source/amazon-kinesis-video-streams-pic/src/utils/src/Mutex.c:202 (libkvsWebrtcClient.so+0x000000060c08)
-#     #2 stepSignalingStateMachine /usr/src/amazon-kinesis-video-streams-webrtc-sdk-c/src/source/Signaling/StateMachine.c:37 (libkvsWebrtcSignalingClient.so+0x000000018c85)
-#     #3 refreshIceConfigurationCallback /usr/src/amazon-kinesis-video-streams-webrtc-sdk-c/src/source/Signaling/Signaling.c:476 (libkvsWebrtcSignalingClient.so+0x00000001d0d7)
-#     #4 timerQueueExecutor /usr/src/amazon-kinesis-video-streams-webrtc-sdk-c/open-source/amazon-kinesis-video-streams-pic/src/utils/src/TimerQueue.c:561 (libkvsWebrtcClient.so+0x000000064c61)
-deadlock:validateIceConfiguration
+race:receiveLwsMe
```
Comment (hassanctech) [inline tst/suppressions/TSAN.supp:113]: Was this intended?  `race:receiveLwsMessage ` --> `race:receiveLwsMe`

File: tst/IceFunctionalityTest.cpp [hunk 3]
```diff
@@ -184,26 +185,33 @@ TEST_F(IceFunctionalityTest, connectionListenerFunctionalityTest)
...
-    EXPECT_EQ(TRUE, IS_VALID_TID_VALUE(pConnectionListener->receiveDataRoutine));
+    // Keeping TSAN happy need to lock/unlock when retrieving the value of TID
+    MUTEX_LOCK(pConnectionListener->lock);
+    threadId = pConnectionListener->receiveDataRoutine;
+    MUTEX_UNLOCK(pConnectionListener->lock);
+    EXPECT_TRUE( IS_VALID_TID_VALUE(threadId));
     ATOMIC_STORE_BOOL(&pConnectionListener->terminate, TRUE);
 
-    THREAD_SLEEP((SOCKET_WAIT_FOR_DATA_TIMEOUT_SECONDS + 1) * HUNDREDS_OF_NANOS_IN_A_SECOND);
+    THREAD_SLEEP(CONNECTION_LISTENER_SOCKET_WAIT_FOR_DATA_TIMEOUT + 1 * HUNDREDS_OF_NANOS_IN_A_SECOND);
```
Comment (hassanctech) [inline tst/IceFunctionalityTest.cpp:209]: Previously sleep time was 2s. Now it is 1.2s.  Is this intended?

Comment (MushMal) [reply tst/IceFunctionalityTest.cpp]: Yeah, my mistake. Fixed

**Review by hassanctech — APPROVED** [review]

**Review by disa6302 — COMMENTED** [review]
Based on our discussion, LGTM.

**Pull request merged**